File: algorithm_practice_python/leetcode/BinarySearch/Sqrt.py
class Sqrt:
    def mySqrt(self, x: int) -> int:
        if x < 2:
            return x

        left = 2
        right = x // 2

        while left <= right:
            mid = (left + right) // 2
            tgt = mid ** 2
            if tgt < x:
                left = mid + 1
            elif tgt > x:
                right = mid - 1
            else:
                return mid

        return right

File: algorithm_practice_java/src/main/java/leetcode/binary_search/Sqrt.java
package leetcode.binary_search;

public class Sqrt {
    public static int mySqrt(int x) {
        if (x < 2) {
            return x;
        }

        int left = 2;
        int right = x / 2;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            long tgt = (long)mid * mid;
            if (tgt < x) {
                left = mid + 1;
            } else if (tgt > x) {
                right = mid - 1;
            } else {
                return mid;
            }
        }

        return right;
    }
}
